packages/darklang/languageServerProtocol/README.md

```diff
@@ -36,13 +36,45 @@ project is quite hard to follow, though, so we've reorganized things here.
 - **notification**: a message sent from one LSP party to another, which does not expect a result
 - **message**: a general term for either a notification, request, or response (error or result)
 
-## Some patterns to follow per-file:
+## Some patterns
 
 - try to keep type names the same as the TypeScript ones in the spec
+- per-file organization
+  - supporting types at the top
+  - types around requests, results, and notifications in the middle
+  - client/server capabilities at the end of the file
+- many types are defined as interfaces with `extends` such as:
 
-- supporting types at the top
-- types around requests, results, and notifications in the middle
-- client/server capabilities at the end of the file
+  ```js
+  export interface CompletionParams extends TextDocumentPositionParams {
+    context?: CompletionContext;
+  }
+  ```
+
+  We could represent this in Darklang by inlining those properties in this type, resulting in something like
+
+  ```fsharp
+  type CompletionParams = {
+    textDocument: TextDocumentIdentifier;
+    position: Position;
+
+    context: CompletionContext;
+  }
+  ```
+
+  or we could add fields per each `extends`:
+
+  ```fsharp
+  type CompletionParams = {
+    textDocumentPosition: TextDocumentPositionParams
+    context: Option<CompletionContext>
+  }
+  ```
+
+  The latter is easier to understand, but won't work out of the box if we end up
+  switching to our usage-ignorant Json parser/serializer. I'm going to use the
+  latter for now, and we can re-evaluate if/when needed. That said, for each case
+  where we do this, we should add a comment roughly of `@extends TextDocumentPositionParams`
```

Comment:
In reflection I think this was a mistake. Going to update the few relevant types to "inline" the relevant fields to match the JSON.